core-bundle/src/EventListener/DataContainer/DefaultOperationsListener.php

```diff
@@ -0,0 +1,222 @@
+        } else {
+            $operations['copy'] = [
+                'href' => 'act=copy',
+                'icon' => 'copy.svg',
+                'button_callback' => $this->isGrantedCallback(CreateAction::class, $table),
+            ];
+        }
+
+        $operations['delete'] = [
+            'href' => 'act=delete',
+            'icon' => 'delete.svg',
+            'attributes' => 'onclick="if(!confirm(\''.($GLOBALS['TL_LANG']['MSC']['deleteConfirm'] ?? null).'\'))return false;Backend.getScrollOffset()"',
+            'button_callback' => $this->isGrantedCallback(DeleteAction::class, $table),
+        ];
+
+        if (null !== ($toggleField = $this->getToggleField($table))) {
+            $operations['toggle'] = [
+                'href' => 'act=toggle&amp;field='.$toggleField,
+                'icon' => 'visible.svg',
+                'button_callback' => $this->isGrantedCallback(UpdateAction::class, $table),
+            ];
+        }
+
+        return $operations + [
+            'show' => [
+                'href' => 'act=show',
+                'icon' => 'show.svg',
+            ],
+        ];
+    }
+
+    private function isGrantedCallback(string $actionClass, string $table): \Closure
+    {
+        return function (DataContainerOperation $operation) use ($actionClass, $table): void {
+            if (!$this->isGranted($actionClass, $table, $operation)) {
+                $this->disableOperation($operation);
+            }
+        };
+    }
+
+    private function copyChildsCallback(string $table): \Closure
+    {
+        return function (DataContainerOperation $operation) use ($table): void {
+            if (!$this->isGranted(CreateAction::class, $table, $operation)) {
+                $this->disableOperation($operation);
+
+                return;
+            }
+
+            $childCount = $this->connection->fetchOne(
+                "SELECT COUNT(*) FROM $table WHERE pid=?",
+                [(string) $operation->getRecord()['id']]
+            );
+
+            if ($childCount < 1) {
+                $this->disableOperation($operation);
+            }
+        };
+    }
+
+    /**
+     * Finds the one and only toggle field in a DCA. Returns null if multiple fields can be toggled.
+     */
+    private function getToggleField(string $table): string|null
+    {
+        $field = null;
+
+        foreach ($GLOBALS['TL_DCA'][$table]['fields'] as $name => $config) {
+            if (!($config['toggle'] ?? false) && !($config['reverseToggle'] ?? false)) {
+                continue;
+            }
+
+            // more than one toggle field exists
+            if (null !== $field) {
+                return null;
+            }
+
+            $field = $name;
+        }
+
+        return $field;
+    }
+
+    private function isGranted(string $actionClass, string $table, DataContainerOperation $operation): bool
+    {
+        $subject = match ($actionClass) {
+            CreateAction::class => new CreateAction($table),
```

Comment:
If the operation has a parent, we have to pass `['pid' => $pid]` here, like we do in DC_Table as well.

Reply:
I just checked on that. The copy (and copychilds) operation check for this permission, but they do not know the new PID yet. We only do that in the core on the _create new after_ automatic operation in DC_Table. We need to add that check to the paste buttons, but that's a different PR since it is a bug in core atm.

Reply:
Can you create a separate issue for this then, so we don't forget about it? Then this is resolved for now :)